server/src/main/java/org/elasticsearch/index/mapper/KeywordFieldMapper.java

```diff
@@ -274,8 +277,33 @@ public TermsEnum getTerms(boolean caseInsensitive, String string, SearchExecutio
             a = MinimizationOperations.minimize(a, Integer.MAX_VALUE);
 
             CompiledAutomaton automaton = new CompiledAutomaton(a);
-            return automaton.getTermsEnum(terms);            
-        }
+            
+            if (automaton.type == AUTOMATON_TYPE.ALL) {
+                TermsEnum result = terms.iterator();
+                if (searchAfter != null) {
+                    result = new SearchAfterTermsEnum(result, searchAfter);
+                }
+                return result;
+            }
+            return terms.intersect(automaton, searchAfter);
+        }
+        
+        // Initialises with a seek to a given term but excludes that term
+        // from results
+        public final class SearchAfterTermsEnum extends FilteredTermsEnum {
```

Comment:
Does this need to be public?

Reply:
I thought I might need it in FlattenedFieldMapper too but didn't in the end. I'll remove the public modifier.